homeassistant/components/here_travel_time/__init__.py

```diff
@@ -1 +1,170 @@
+"""The HERE Travel Time integration."""
+from __future__ import annotations
+
+from datetime import datetime, time, timedelta
+import logging
+
+import async_timeout
+from herepy import NoRouteFoundError, RouteMode, RoutingApi, RoutingResponse
+
+from homeassistant.const import ATTR_ATTRIBUTION, CONF_UNIT_SYSTEM_IMPERIAL, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.location import find_coordinates
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt
+
+from .const import (
+    ATTR_DESTINATION,
+    ATTR_DESTINATION_NAME,
+    ATTR_DISTANCE,
+    ATTR_DURATION,
+    ATTR_DURATION_IN_TRAFFIC,
+    ATTR_ORIGIN,
+    ATTR_ORIGIN_NAME,
+    ATTR_ROUTE,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    NO_ROUTE_ERROR_MESSAGE,
+    TRAFFIC_MODE_ENABLED,
+    TRAVEL_MODES_VEHICLE,
+    HERERoutingData,
+    HERETravelTimeConfig,
+)
+
+PLATFORMS = [Platform.SENSOR]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HereTravelTimeDataUpdateCoordinator(DataUpdateCoordinator):
+    """HERETravelTime DataUpdateCoordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        api: RoutingApi,
+        config: HERETravelTimeConfig,
+    ) -> None:
+        """Initialize."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=DEFAULT_SCAN_INTERVAL),
+        )
+        self._api = api
+        self.config = config
+
+    async def _async_update_data(self) -> HERERoutingData | None:
+        """Get the latest data from the HERE Routing API."""
+        try:
+            async with async_timeout.timeout(10):
+                return await self.hass.async_add_executor_job(self._update)
+        except NoRouteFoundError as error:
+            raise UpdateFailed(NO_ROUTE_ERROR_MESSAGE) from error
+
+    def _update(self) -> HERERoutingData | None:
+        """Get the latest data from the HERE Routing API."""
+        if self.config.origin_entity_id is not None:
+            origin = find_coordinates(self.hass, self.config.origin_entity_id)
+        else:
+            origin = self.config.origin
+
+        if self.config.destination_entity_id is not None:
+            destination = find_coordinates(self.hass, self.config.destination_entity_id)
+        else:
+            destination = self.config.destination
+        if destination is not None and origin is not None:
+            here_formatted_destination = destination.split(",")
+            here_formatted_origin = origin.split(",")
+            arrival: str | time | None = None
+            departure: str | time | None = "now"
+            if (arrival := self.config.arrival) is not None:
+                arrival = convert_time_to_isodate(str(arrival))
+            if (departure := self.config.departure) is not None:
+                departure = convert_time_to_isodate(str(departure))
```

Comment:
I realize you're wanting to not change functions, etc, but given all the typing that was introduced here and the configuration `arrival` is now typed as a `time` I think the conversion from time => string => time conversion should be removed and simplified so we can do something like this:

```suggestion
            arrival: str | None = None
            departure: str | None = "now"
            if self.config.arrival is not None:
                arrival = convert_time_to_isodate(self.config.arrival)
            if self.config.departure is not None:
                departure = convert_time_to_isodate(self.config.departure)
```